Mathlib/CategoryTheory/Sites/LocallyInjective.lean

```diff
@@ -0,0 +1,196 @@
+
+instance [IsIso φ] : IsLocallyInjective J φ :=
+  isLocallyInjective_of_injective J φ (fun X => Function.Bijective.injective (by
+    rw [← isIso_iff_bijective]
+    change IsIso ((forget D).map (φ.app X))
+    infer_instance))
+
+instance isLocallyInjective_forget [IsLocallyInjective J φ] :
+    IsLocallyInjective J (whiskerRight φ (forget D)) where
+  equalizerSieve_mem x y h := equalizerSieve_mem J φ x y h
+
+instance isLocallyInjective_comp [IsLocallyInjective J φ] [IsLocallyInjective J ψ] :
+    IsLocallyInjective J (φ ≫ ψ) where
+  equalizerSieve_mem {X} x y h := by
+    let S := equalizerSieve (φ.app _ x) (φ.app _ y)
+    let T : ∀ ⦃Y : C⦄ ⦃f : Y ⟶ X.unop⦄ (_ : S f), Sieve Y := fun Y f _ =>
+      equalizerSieve (F₁.map f.op x) ((F₁.map f.op y))
+    refine J.superset_covering ?_
+      (J.transitive (equalizerSieve_mem J ψ (φ.app _ x) (φ.app _ y) (by simpa using h))
+      (Sieve.bind S.1 T) ?_)
+    · rintro Y f ⟨Z, a, g, hg, ha, rfl⟩
+      simpa using ha
+    · intro U f hf
+      exact J.superset_covering (Sieve.le_pullback_bind S.1 T _ hf)
+        (equalizerSieve_mem J φ (F₁.map f.op x) (F₁.map f.op y) (by simpa using hf))
+
+lemma isLocallyInjective_of_isLocallyInjective [IsLocallyInjective J (φ ≫ ψ)] :
+    IsLocallyInjective J φ where
+  equalizerSieve_mem {X} x y h := equalizerSieve_mem J (φ ≫ ψ) x y (by simp [h])
+
+variable {φ ψ}
+
+lemma isLocallyInjective_of_isLocallyInjective_fac {φψ : F₁ ⟶ F₃} (fac : φ ≫ ψ = φψ)
+    [IsLocallyInjective J φψ] : IsLocallyInjective J φ := by
+  subst fac
+  exact isLocallyInjective_of_isLocallyInjective J φ ψ
+
+lemma isLocallyInjective_iff_of_fac {φψ : F₁ ⟶ F₃} (fac : φ ≫ ψ = φψ) [IsLocallyInjective J ψ] :
+    IsLocallyInjective J φψ ↔ IsLocallyInjective J φ := by
+  constructor
+  · intro
+    exact isLocallyInjective_of_isLocallyInjective_fac J fac
+  · intro
+    rw [← fac]
+    infer_instance
+
+variable (φ ψ)
+
+lemma isLocallyInjective_comp_iff [IsLocallyInjective J ψ] :
+    IsLocallyInjective J (φ ≫ ψ) ↔ IsLocallyInjective J φ :=
+  isLocallyInjective_iff_of_fac J rfl
+
+lemma isLocallyInjective_iff_injective_of_separated (hsep : IsSeparated J F₁) :
+    IsLocallyInjective J φ ↔ ∀ (X : Cᵒᵖ), Function.Injective (φ.app X) := by
+  constructor
+  · intro _ X x y h
+    exact hsep X.unop _ (equalizerSieve_mem J φ x y h) _ _ (fun _ _ hf => hf)
+  · apply isLocallyInjective_of_injective
+
+instance (F : Cᵒᵖ ⥤ Type w) (G : GrothendieckTopology.Subpresheaf F) :
+    IsLocallyInjective J G.ι :=
+  isLocallyInjective_of_injective _ _ (fun X => by
+    intro ⟨x, _⟩ ⟨y, _⟩ h
+    exact Subtype.ext h)
+
+section
+
+variable {E : Type u'} [Category.{max u v} E] [ConcreteCategory E]
+  [PreservesLimits (forget E)]
+  [∀ (P : Cᵒᵖ ⥤ E) (X : C) (S : J.Cover X),
+    HasMultiequalizer (GrothendieckTopology.Cover.index S P)]
+  [∀ (X : C), HasColimitsOfShape (GrothendieckTopology.Cover J X)ᵒᵖ E]
+  [∀ X : C, PreservesColimitsOfShape (J.Cover X)ᵒᵖ (forget E)] [(forget E).ReflectsIsomorphisms]
```

Comment:
Is it possible to use `[ConcreteCategory E] [HasSheafify J E] [HasSheafCompose J (forget E)]` instead?

Reply:
Unfortunately, I do not know how to prove that the map to the associated sheaf is locally injective without using its explicit description (which I am doing in two steps using the plus construction).

Reply:
I thought those conditions guarantee that `forget E` preserves sheafification, but maybe a new class
```lean
class PreservesSheafification {D E} [Category D] [Category E]
    [HasWeakSheafify J D] [HasWeakSheafify J E] (F : D ⥤ E) where
  iso : ∀ P : Cᵒᵖ ⥤ D, sheafify J P ⋙ F ≅ sheafify J (P ⋙ F)
  prop : ∀ P : Cᵒᵖ ⥤ D, whiskerRight (toSheafify J P) F ≫ (iso P).hom = toSheafify J (P ⋙ F)
```
is needed here.
I think the main purpose of these abstract classes is to hide the plus plus construction as an implementation detail and I think we should discourage the use of the latter outside of the construction files if possible.
 
This can come in a later PR though.

Reply:
Actually, what would be also useful (after developing a little bit more locally surjective/injective maps), is to introduce a type-class saying that a morphism of presheaves (in a concrete category) is in `J.W` (i.e. iso after sheafification) iff it is both locally injective and locally surjective. At some point, we have to show that under certain circumstances, this holds, and the code in this PR shall help doing that. (In the future, I plan to allow sharper universe bounds using certain PR on 1-hypercovers, like #13012.)

Reply:
This section can be replaced by 
(`infer_instance` is able to prove the old `isLocallyInjective_toSheafify'` from the new one)
```
variable (P : Cᵒᵖ ⥤ Type max u v)

open GrothendieckTopology Plus

instance isLocallyInjective_toPlus : IsLocallyInjective J (J.toPlus P) where
  equalizerSieve_mem {X} x y h := by
    erw [toPlus_eq_mk, toPlus_eq_mk, eq_mk_iff_exists] at h
    obtain ⟨W, h₁, h₂, eq⟩ := h
    exact J.superset_covering (fun Y f hf => congr_fun (congr_arg Subtype.val eq) ⟨Y, f, hf⟩) W.2

instance isLocallyInjective_toSheafify : IsLocallyInjective J (J.toSheafify P) := by
  dsimp [GrothendieckTopology.toSheafify]
  rw [GrothendieckTopology.plusMap_toPlus]
  infer_instance

instance isLocallyInjective_toSheafify' [ConcreteCategory.{max u v} D]
    (P : Cᵒᵖ ⥤ D) [HasWeakSheafify J D] [J.HasSheafCompose (forget D)]
    [J.PreservesSheafification (forget D)] :
    IsLocallyInjective J (toSheafify J P) := by
  rw [← isLocallyInjective_forget_iff, ← sheafComposeIso_hom_fac,
    ← toSheafify_plusPlusIsoSheafify_hom]
  infer_instance

end
```

Reply:
Thanks for this code!